File: python/playwright-async-file-download/scrape.py
#!/usr/bin/env python3
import asyncio
from os import environ
from base64 import standard_b64decode
from playwright.async_api import Playwright, Page, CDPSession, async_playwright

AUTH = environ.get('AUTH', default='USER:PASS')
TARGET_URL = environ.get('TARGET_URL',
                         default='https://myjob.page/tools/test-files')
SELECTOR = environ.get('SELECTOR', default='a[role=button]')
FILENAME = environ.get('FILENAME', default='./testfile.zip')


async def scrape(playwright: Playwright, url=TARGET_URL,
                 selector=SELECTOR, filename=FILENAME):
    if AUTH == 'USER:PASS':
        raise Exception('Provide Scraping Browsers credentials in AUTH ' +
                        'environment variable or update the script.')
    print('Connecting to Browser...')
    endpoint_url = f'wss://{AUTH}@brd.superproxy.io:9222'
    browser = await playwright.chromium.connect_over_cdp(endpoint_url)
    try:
        print(f'Connected! Navigating to {url}...')
        page = await browser.new_page()
        client = await page.context.new_cdp_session(page)
        await page.goto(url, timeout=2*60_000)
        print('Navigated! Initiating download...')
        request_id = await initiate_download(page, client, selector)
        print(f'Download started! Stream it to {filename}...')
        stream = await client.send('Fetch.takeResponseBodyAsStream', {
            'requestId': request_id,
        })
        file = open(filename, 'wb')
        file_size = 0
        while True:
            chunk = await client.send('IO.read', {
                'handle': stream['stream'],
            })
            if chunk['base64Encoded']:
                data = standard_b64decode(chunk['data'])
            else:
                data = bytes(chunk['data'], 'utf8')
            file.write(data)
            file_size += len(data)
            if chunk['eof']:
                break
        file.close()
        print(f'Download saved! Size: {file_size}')
    finally:
        await browser.close()


async def initiate_download(page: Page, client: CDPSession, selector):
    future = asyncio.get_running_loop().create_future()

    def on_request_paused(event):
        request_id = event['requestId']
        if future.done():
            asyncio.ensure_future(client.send('Fetch.continueRequest', {'requestId': request_id}))
        else:
            future.set_result(request_id)

    def on_click(click: asyncio.Future):
        if not future.done() and click.exception():
            future.set_exception(click.exception())

    await client.send('Fetch.enable', {
        'patterns': [{
            'requestStage': 'Response',
            'resourceType': 'Document',
        }],
    })
    client.on('Fetch.requestPaused', on_request_paused)
    asyncio.ensure_future(page.click(selector)).add_done_callback(on_click)
    return await future


async def main():
    async with async_playwright() as playwright:
        await scrape(playwright)


if __name__ == '__main__':
    asyncio.run(main())

File: nodejs/playwright-file-download/scrape.js
#!/usr/bin/env node
const playwright = require('playwright');
const fs = require('fs/promises');
const {
    AUTH = 'USER:PASS',
    TARGET_URL = 'https://calmcode.io/datasets/bigmac',
    SELECTOR = 'button.border',
    FILENAME = './testfile.csv',
} = process.env;

async function scrape(url = TARGET_URL, selector = SELECTOR, filename = FILENAME) {
    if (AUTH == 'USER:PASS') {
        throw new Error(`Provide Scraping Browsers credentials in AUTH`
            + ` environment variable or update the script.`);
    }
    console.log(`Connecting to Browser...`);
    const endpointURL = `wss://${AUTH}@brd.superproxy.io:9222`;
    const browser = await playwright.chromium.connectOverCDP(endpointURL);
    try {
        console.log(`Connected! Navigating to ${url}...`);
        const page = await browser.newPage();
        const client = await page.context().newCDPSession(page);
        await page.goto(url, { timeout: 2 * 60 * 1000 });
        console.log(`Navigated! Initiating download...`);
        const requestId = await initiateDownload(page, client, selector);
        console.log(`Download started! Stream it to ${filename}...`);
        const { stream } = await client.send('Fetch.takeResponseBodyAsStream', {
            requestId,
        });
        const file = await fs.open(filename, 'w');
        let fileSize = 0;
        while (true) {
            const { data, base64Encoded, eof } = await client.send('IO.read', {
                handle: stream,
            });
            const chunk = Buffer.from(data, base64Encoded ? 'base64' : 'utf8');
            await file.write(chunk);
            fileSize += chunk.byteLength;
            if (eof) {
                break;
            }
        }
        await file.close();
        console.log(`Download saved! Size: ${fileSize}.`);
    } finally {
        await browser.close();
    }
}

async function initiateDownload(page, client, selector) {
    await client.send('Fetch.enable', {
        patterns: [{
            requestStage: 'Response',
            resourceType: 'Document',
        }],
    });
    return await new Promise((resolve, reject) => {
        client.on('Fetch.requestPaused', ({ requestId }) => {
            resolve(requestId);
        });
        page.click(selector).catch(error => {
            reject(error);
        });
    });
}

if (require.main == module) {
    scrape().catch(error => {
        console.error(error.stack || error.message || error);
        process.exit(1);
    });
}

